File: 5-Problem-A-Day/89. Gray Code.java
class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> ans = new LinkedList<>();
        if (n >= 0)
            ans.add(0);
        if (n >= 1)
            ans.add(1);
        for (int i = 2; i <= n; i++) {
            List<Integer> tmp = new LinkedList<>(ans);
            for (int j = ans.size() - 1; j > -1; j--) {
                tmp.add(ans.get(j) | (1 << (i - 1)));
            }
            ans = new LinkedList<>(tmp);
        }
        return ans;
    }
}

File: 2023/89. Gray Code.cpp
class Solution {
public:
    vector<int> grayCode(int n) {
        if (n == 1) {
            return {0, 1};
        }
        vector<int> prev = grayCode(n - 1);
        vector<int> res;
        for (int v : prev) {
            res.push_back((v << 1) + 0);
        }
        for (int i = prev.size() - 1; i >= 0; i--) {
            res.push_back((prev[i] << 1) + 1);
        }
        return res;
    }
};